### PR TITLE
ENH: Modified stats.kstest() to use the exact stats.kstwo.sf() where appropriate

### PR DESCRIPTION
**Pull request by pvanmulbregt**
#### Reference issue
gh-9797  Master tracker for K-S Issues. 

#### What does this implement/fix?

This is the third and final in a series of PRs (gh-9719, gh-9753) rewriting
  the Kolmogorov-Smirnov test code.
Add `'exact'`, `'auto'` as values for `mode` in `kstest(...,mode=)`:
  `'exact'` results in calling `stats.kstwo.sf()` instead of the asymptotic `stats.kstwobign.sf()`
For `alternative='two-sided'`,  `'auto'` uses `'exact'` if N is not too large,
  otherwise it uses `'asymp'`.
Change the default value of `mode` from `'asymp'` to `'auto'`.  This results in
  different, but more accurate, values being returned by `stats.kstest()` if the
  `mode` is not specified.
Change `mstats_basic.ks_twosamp()` to call `stats.ks_2samp()` instead of
  performing any calculations itself.
Update returned values in the tutorial and docstrings.

### CODE REVIEW

**Comment by pvanmulbregt** [comment]
The one Travis CI failure is due to gh-11711.

File: scipy/stats/stats.py [hunk 1]
```diff
@@ -5297,12 +5329,14 @@ def kstest(rvs, cdf, args=(), N=20, alternative='two-sided', mode='approx'):
...
+    mode : {'auto', 'exact', 'approx', 'asymp'}, optional
         Defines the distribution used for calculating the p-value.
-        The following options are available (default is 'approx'):
+        The following options are available (default is 'auto'):
 
-          * 'approx': use approximation to exact distribution of test statistic
-          * 'asymp': use asymptotic distribution of test statistic
+          * 'auto' : uses 'exact' for small size arrays, 'asymp' for large.
+          * 'exact' : uses the exact distribution of test statistic.
+          * 'approx' : approximates the two-sided probability with twice the one-sided probability
```
Comment (chrisb83) [inline scipy/stats/stats.py:5338]: what is the use case of approx?

File: scipy/stats/stats.py [hunk 2]
```diff
@@ -5297,12 +5329,14 @@ def kstest(rvs, cdf, args=(), N=20, alternative='two-sided', mode='approx'):
           * 'two-sided'
           * 'less': one-sided, see explanation in Notes
           * 'greater': one-sided, see explanation in Notes
-    mode : {'approx', 'asymp'}, optional
+    mode : {'auto', 'exact', 'approx', 'asymp'}, optional
         Defines the distribution used for calculating the p-value.
-        The following options are available (default is 'approx'):
+        The following options are available (default is 'auto'):
```
Comment (chrisb83) [inline scipy/stats/stats.py:5334]: is the mode only relevant for 'two-sided'?

File: scipy/stats/stats.py [hunk 3]
```diff
@@ -5378,48 +5412,48 @@ def kstest(rvs, cdf, args=(), N=20, alternative='two-sided', mode='approx'):
...
+    if alternative == 'less':
+        Dminus = _compute_dminus(cdfvals)
+        return KstestResult(Dminus, distributions.ksone.sf(Dminus, N))
+
+    # alternative == 'two-sided':
+    Dplus = _compute_dplus(cdfvals)
+    Dminus = _compute_dminus(cdfvals)
+    D = np.max([Dplus, Dminus])
+    if mode == 'auto':
+        mode = 'exact' if N < 1000000 else 'asymp'
```
Comment (chrisb83) [inline scipy/stats/stats.py:5447]: 1e6 seems to be very large, wouldn't the asymptotic formula work well for much smaller N, e.g. N=1000?

File: scipy/stats/stats.py [hunk 4]
```diff
@@ -5378,48 +5412,48 @@ def kstest(rvs, cdf, args=(), N=20, alternative='two-sided', mode='approx'):
...
+    if mode == 'exact':
+        prob = distributions.kstwo.sf(D, N)
+    elif mode == 'asymp':
+        prob = distributions.kstwobign.sf(D * np.sqrt(N))
+    else:
+        # mode == 'approx'
+        prob = 2 * distributions.ksone.sf(D, N)
```
Comment (chrisb83) [inline scipy/stats/stats.py:5454]: do you know if the old logic including the check `N > 2666 or pval_two > 0.80 - N*0.3/1000` offers any advantage?

**Review by chrisb83 — COMMENTED** [review]
Hi Paul, 

I left a few comments to clarify some points.

~~For the mstats version, should we consider adding the keyword mode?~~ [DONE]

~~Could you also link the tests epps_singleton_2samp and anderson_ksamp in the see also section?~~[DONE]

Thanks

File: scipy/stats/mstats_basic.py [hunk 5]
```diff
@@ -1282,30 +1282,12 @@ def ks_twosamp(data1, data2, alternative="two-sided"):
...
-    elif alternative == 'g':
-        d = csum.max()
-        prob = np.exp(-2*n*d**2)
-    else:
-        raise ValueError("Invalid value for the alternative hypothesis: "
-                         "should be in 'two-sided', 'less' or 'greater'")
-
+    alternative = {'t': 'two-sided', 'g':'greater', 'l': 'less'}.get(
+       alternative.lower()[0], alternative)
+    data1 = ma.asarray(data1)
```
[the remark this answers is not in the document]
Comment (pvanmulbregt) [reply scipy/stats/mstats_basic.py]: Appended the `compressed()` calls onto the `asarray()` calls.

File: scipy/stats/stats.py (hunk 1 above)
Comment (pvanmulbregt) [reply scipy/stats/stats.py:5338]: `approx` used to return either the asymptotic value `kstwobign.sf(D * np.sqrt(N))`, or twice the one-sided probability `2 * ksone.sf(D, N)`, depending on whether `N > 2666` and other criteria which you asked about below.
Now `approx` always means `2 * ksone.sf(D, N)`.
`auto` will choose one of  [`exact` or `asymp`], never `approx`.
The retention of `approx` is required for backwards compatibility.
I'd expect new  code calling `kstest()` to use the default `mode=auto`.

File: scipy/stats/stats.py (hunk 2 above)
Comment (pvanmulbregt) [reply scipy/stats/stats.py:5334]: Yes.  The one-sided tests call  `distributions.ksone.sf` and that decides when to switch from exact to an approximation.

File: scipy/stats/stats.py (hunk 3 above)
Comment (pvanmulbregt) [reply scipy/stats/stats.py]: Upon reflection, it's not clear to me whether this switchover needs to exist any more.  The `exact` path calls `kstwo.sf(D, N)` which knows when to approximate with the asymptotic.  The asymptotic likely matches well for large `N`, except for small `D`, those in the range `[0, 2/N]` say, as the exact value approaches `1` faster than the asymptotic.

File: scipy/stats/stats.py (hunk 4 above)
Comment (pvanmulbregt) [reply scipy/stats/stats.py:5455]: That seems to have been a heuristic approximation whose origin is unknown to me (no reference in the code, no mention in any papers I've seen, and IIRC Josef didn't know either.).  Certainly it is less accurate.

File: scipy/stats/stats.py (hunk 1 above)
Comment (chrisb83) [reply scipy/stats/stats.py:5338]: OK, maybe add a comment to the docstring to clarify that this option is mainly for backwards compatability and it is not needed in most situations (maybe some users will test how well the approximation works if they are familiar with the test). The approx option is not available in mstats for that reason I assume

File: scipy/stats/stats.py (hunk 2 above)
Comment (chrisb83) [reply scipy/stats/stats.py:5334]: should we mention this in the documentation?

File: scipy/stats/stats.py (hunk 3 above)
Comment (chrisb83) [reply scipy/stats/stats.py]: agree, one could even consider to use 'exact' as the default option?

File: scipy/stats/stats.py (hunk 4 above)
Comment (chrisb83) [reply scipy/stats/stats.py:5455]: ok, it is better to clean up the implementation in that case

File: scipy/stats/__init__.py [hunk 6]
```diff
@@ -263,6 +263,7 @@
    brunnermunzel
    combine_pvalues
    jarque_bera
+   Ks_2sampResult
```
Comment (chrisb83) [inline scipy/stats/__init__.py:266]: for the other tests that return named tulples, we do not mention the output here. do you propose to change that?

**Comment by pvanmulbregt** [comment]
The latest push had quite a few changes so I'll put the description here and then add some comments.

--------------
ENH: Modified stats.kstest() to use stats.kstwo.sf().

This is the third and final in a series of PRs (#9719, #9753) rewriting
  the Kolmogorov-Smirnov test code.
Add 'exact', 'auto' as values for mode in stats.kstest(...,mode=):
  'exact' calls stats.kstwo.sf() instead of stats.kstwobign.sf()
For alternative='two-sided',  'auto' uses 'exact' if N is not too large,
  otherwise it uses 'asymp'.
Change the default value of mode from 'asymp' to 'auto'.  This results in
  different, but more accurate, values being returned by stats.kstest() if the
  mode is not specified.
Enhance stats.kstest(data1, data2, ...) so that it performs the 1-sample test if
  given a cdf function, and the 2=sample test if provided two arrays.
Support masked arrays in stats.{kstest,ks_1samp,ks_2samp}.  Previously masked
  arrays were accepted but the wrong answer was returned.
Change mstats_basic.ks_twosamp() to call stats.ks_2samp() instead of
  performing any calculations itself.
Added mstats_basic.ks_2samp() to call stats.ks_2samp().
Added mstats_basic.kstest() to call stats.kstest().
Update returned values in the tutorial and docstrings.
Updated test values in (@pytest.mark.slow) TestKSTwoSamples.testLargeBoth().
Reduced exact-to-asymp switchover from 1000000 to 10000 in stats.kstest().
Added test of stats.kstest()'s multiple code paths in kstest: test_kstest_allpaths().
Added TestCompareWithStats.test_ks_2samp() to check stats/mstats agreement.
Added handling of NaN values in _kstest.py's API routines.
Replaced a manually computed product with np.prod(expr) in stats.ks_2samp().
Added some comments in a stats test function: test_kstest_allpaths().
Moved kstest and ks_2samp next to each other in scipy/stats/stats.py.
Moved _contains_nan from mstats_basic.py to stats.py.
Add tests of the new functionality.

**Comment by pvanmulbregt** [comment]
1. The basic premise remains the same: Have K-S testing call the appropriate `stats.distribution.{ksone|kstwo|kstwobign}.cdf` to calculate the p-value of the K-S statistic. 


1. While addressing @chrisb83 's comments I noticed that while `stats.mstats_basic` has some support for KS testing using masked arrays, the basic `stats.stats` implementation of `kstest` also (silently) accepted masked arrays.   It just shunted all the masked values to the end which produced the correct statistic but the the wrong p-value. I added support for masked arrays into  the `stats.stats` KS functions and have the `mstats_basic` functions call the `stats` versions.  This also makes  `mstats_basic` and `stats` kstests agree _a priori_, rather than having to keep them in sync.

1. I also enhanced the signature of `kstest()` to support both one sample and two sample testing.
```
def kstest(data1, data2, args=(), N=20, alternative='two-sided', mode='auto'):
```
`kstest` now decodes the argument possibilities and dispatches to either the one- or two-sample test.

 - `kstest(samples1, samples2,...)` => two-sample testing 
 - `kstest(samples1, cdffunc, args=...)` => one-sample testing, against `cdffunc(*args)`
 - `kstest(samples1, "norm", args=...)` => one-sample testing, against `stats.norm.cdf(*args)` [`"norm"` can be replaced with the name of any distribution in `stats.distributions`.]
 - `kstest(rvsfunc, cdffunc, args=..., N=...)` => one-sample testing, compute a sample of size N using `rvsfunc(*args, size=N)` against `cdffunc(*args)`
 - `kstest("norm", cdffunc, args=..., N=...)` => one-sample testing, compute a sample of size N using `stats.distributions.norm.rvs(*args, size=N)`, against `cdffunc(*args)`. 


1. There are some unusual, all equivalent, additional possibilities which date back to a commit in 2008 3e3e8eaef5ff26148c3f79c925b4d20b15e05988.

 - `kstest("norm", "norm", args=..., N=...)` => one-sample testing, compute a sample of size N using `distributions.norm.rvs(*args, size=N)`, against `distributions.norm.cdf(*args)`
 - `kstest("norm", "", args=..., N=...)` => same
 - `kstest("norm", False, args=..., N=...)` => same
 - `kstest("norm", [], args=..., N=...)` => same
 - `kstest("norm", 0, args=..., N=...)` => same
 -  In fact, if `data1` is a string then any value of `data2` that evaluates False has this result.

What is the purpose of this variation, telling `kstest` to generate a random sample from a distribution and then comparing the sample against that same distribution's CDF?  Is it  just for some ease of testing or is there a real-world use case? @josef-pkt perhaps you have some context here?

File: scipy/stats/__init__.py (hunk 6 above)
Comment (pvanmulbregt) [reply scipy/stats/__init__.py]: By having `mstats_basic` K-S functions call `stats` K-S functions, I no longer needed to expose that named tuple. 
Exposing them all in the API would require taking a look at the names and rationalizing them, making consistent, as that is not the case, from a quick perusal.

**Comment by pvanmulbregt** [comment]
Closes gh-11884 as a side-effect.

**Comment by pvanmulbregt** [comment]
The `kstest, ks_2samp, ks_1samp` family of functions work for both masked arrays and non-masked arrays. Found a way to import the functions from `stats` into the `mstats` namespace, needed for backward compatibility. (`mstats_basic` and `stats` are unfortunately circularly dependent.)

**Comment by pvanmulbregt** [comment]
Spoke too soon.  Importing `ks_2samp` from `stats` into `mstats` requires that any references in the docstring are also visible or the doc builds fail. will restore the shims in `mstats`.

File: scipy/stats/stats.py (hunk 3 above)
Comment (chrisb83) [reply scipy/stats/stats.py]: the current version is:
```if mode == 'auto':  # Always select exact
    mode = 'exact'
```
i think auto is not even needed for backwards compatibility since it is introduced in this PR. should we go with exact as the default option or do I miss something?

**Comment by chrisb83** [comment]
@pvanmulbregt let me know if I should continue my review. (unfortunately the diff of `stats.py` is displayed in a way that is very hard to read, some functions are shown as deleted and reappear somewhere else.)

**Comment by pvanmulbregt** [comment]
>  let me know if I should continue my review. (unfortunately the diff of `stats.py` is displayed in a way that is very hard to read, some functions are shown as deleted and reappear somewhere else.)

@chrisb83 Yes, it is fine to continue.  (The strange diff in stats may be because I moved the `kstest` function below the worker functions, and it contains the docstrings.)

File: scipy/stats/stats.py [hunk 7]
```diff
@@ -5819,7 +5689,185 @@ def chisquare(f_obs, f_exp=None, ddof=0, axis=0):
...
+    >>> x = stats.norm.rvs(loc=0.2, size=100)
+    >>> stats.kstest(x,'norm', alternative = 'less')
+    (0.12464329735846891, 0.040989164077641749)
+
+    Reject equal distribution against alternative hypothesis: less
+
+    >>> stats.kstest(x,'norm', alternative = 'greater')
+    (0.0072115233216311081, 0.98531158590396395)
+
+    Don't reject equal distribution against alternative hypothesis: greater
```
Comment (chrisb83) [inline scipy/stats/stats.py:5811]: I did not understand this comment. 

Why is asymp used below?

File: scipy/stats/stats.py [hunk 8]
```diff
@@ -5819,7 +5689,185 @@ def chisquare(f_obs, f_exp=None, ddof=0, axis=0):
...
+
+    >>> stats.kstest(x,'norm', mode='asymp')
```
[the remark this answers is not in the document]
Comment (chrisb83) [reply scipy/stats/stats.py:5813]: there are a few missing spaces below

File: scipy/stats/stats.py [hunk 9]
```diff
@@ -5819,7 +5689,185 @@ def chisquare(f_obs, f_exp=None, ddof=0, axis=0):
...
+    (0.12464329735846891, 0.08944488871182088)
+
+    *Testing t distributed random variables against normal distribution*
```
Comment (chrisb83) [inline scipy/stats/stats.py:5816]: the example below is much easier to follow than the one above that has less comments on the idea behind the tests

**Review by chrisb83 — COMMENTED** [review]
Hi Paul,

some comments, mainly on the documentation of ks_1samp in stats

**Comment by chrisb83** [comment]
one more observation: in mstats, `ks_twosamp` is be a candidate for deprecation given that there is already `ks_2samp`?

File: scipy/stats/stats.py [hunk 10]
```diff
@@ -6183,71 +6284,197 @@ def ks_2samp(data1, data2, alternative='two-sided', mode='auto'):
...
+    # kstest allows many different variations of arguments.
+    # Pull out the parsing into a separate function
+    # (xvals, yvals, )  # 2sample
+    # (xvals, cdf function,..)
+    # (xvals, name of distribution, ...)
+    # (name of distribution, name of distribution, ...)
+    # (name of distribution, empty string, ...)
+    # (name of distribution, False, ...)
+
+    # Returns xvals, yvals, cdf
```
Comment (chrisb83) [inline scipy/stats/stats.py:6321]: which combinations are possible for xvals?

**Comment by chrisb83** [comment]
> kstest now decodes the argument possibilities and dispatches to either the one- or two-sample test. 
kstest(samples1, samples2,...) => two-sample testing
kstest(samples1, cdffunc, args=...) => one-sample testing, against cdffunc(*args)
kstest(samples1, "norm", args=...) => one-sample testing, against stats.norm.cdf(*args)

The above use case are the normal ones in my view. The other combinations make the API very complicated and hard to follow and from the current documentation, I am not sure if the behaviour of things like `kstest("norm", False, args=..., N=...)` are clear.

In any case, since this PR just builds on the exising API, so I am fine with allowing such cases.

File: scipy/stats/_ksstats.py [hunk 11]
```diff
@@ -414,25 +414,20 @@ def _kolmogn(n, x, cdf=True):
...
-    # Compute the CDF and take its complement
-    cdfprob = _kolmogn(n, x, cdf=True)
-    return _clip_prob(1.0 - cdfprob)
+    if not cdf:
+        if nxsquared >= 370.0:
+            return 0.0
+        if nxsquared >= 2.2:
+            prob = 2 * scipy.special.smirnov(n, x)
+            return _clip_prob(prob)
+        # Fall through and compute 1-cdf
```
[the remark this answers is not in the document]
Comment (pvanmulbregt) [reply scipy/stats/_ksstats.py:423]: Should be "Compute SF as 1-CDF."

File: scipy/stats/_ksstats.py [hunk 12]
```diff
@@ -414,25 +414,20 @@ def _kolmogn(n, x, cdf=True):
...
+    if nxsquared >= 18.0:
+        cdfprob = 1.0
+    elif n <= 100000 and n * x**1.5 <= 1.4:
+        cdfprob = _kolmogn_DMTW(n, x, cdf=True)
+    else:
+        cdfprob = _kolmogn_PelzGood(n, x, cdf=True)
+    return _select_and_clip_prob(cdfprob, 1.0 - cdfprob, cdf=cdf)
```
[the remark this answers is not in the document]
Comment (pvanmulbregt) [reply scipy/stats/_ksstats.py:430]: Not quite.  If `n > 140` and `n*x*x < 2.2`, the SF  is computed as 1-CDF.

File: scipy/stats/stats.py [hunk 13]
```diff
@@ -5819,7 +5689,185 @@ def chisquare(f_obs, f_exp=None, ddof=0, axis=0):
...
+    This performs a test of the distribution F(x) of an observed
+    random variable against a given distribution G(x). Under the null
+    hypothesis, the two distributions are identical, F(x)=G(x). The
+    alternative hypothesis can be either 'two-sided' (default), 'less'
+    or 'greater'. The KS test is only valid for continuous distributions.
+
+    Parameters
+    ----------
+    x : array_like
+        a 1-D array of observations of random variables.
```
[the remark this answers is not in the document]
Comment (pvanmulbregt) [reply scipy/stats/stats.py:5740]: Done.

File: scipy/stats/stats.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (pvanmulbregt) [reply scipy/stats/stats.py:5813]: Done.

**Comment by pvanmulbregt** [comment]
> > kstest now decodes the argument possibilities and dispatches to either the one- or two-sample test.
> > kstest(samples1, samples2,...) => two-sample testing
> > kstest(samples1, cdffunc, args=...) => one-sample testing, against cdffunc(*args)
> > kstest(samples1, "norm", args=...) => one-sample testing, against stats.norm.cdf(*args)
> 
> The above use case are the normal ones in my view. The other combinations make the API very complicated and hard to follow and from the current documentation, I am not sure if the behaviour of things like `kstest("norm", False, args=..., N=...)` are clear.
> 
> In any case, since this PR just builds on the exising API, so I am fine with allowing such cases.

Personally I'd be in favor of dropping support for these cases.  Nobody jumped to explain the use cases after https://github.com/scipy/scipy/pull/11710#issuecomment-616149862 and they do make the API complicated and difficult to maintain.

File: scipy/stats/stats.py (hunk 10 above)
Comment (pvanmulbregt) [reply scipy/stats/stats.py:6321]: xvals is array_like.  Updated documentation.

**Comment by pvanmulbregt** [comment]
> one more observation: in mstats, `ks_twosamp` is a candidate for deprecation given that there is already `ks_2samp`?

`ks_twosamp` is in the API.  But it would make sense to deprecate the `XX_twosamp` routines. E.g. `ks_twosamp` in `stats` and `mstats`, `ttest_onesamp` in `mstats`.

File: scipy/stats/stats.py [hunk 14]
```diff
@@ -5819,7 +5689,185 @@ def chisquare(f_obs, f_exp=None, ddof=0, axis=0):
...
+
+          * 'auto' : selects one of the other options.
+          * 'exact' : uses the exact distribution of test statistic.
+          * 'approx' : approximates the two-sided probability with twice the one-sided probability
+          * 'asymp': uses asymptotic distribution of test statistic
+
+    Returns
+    -------
+    statistic : float
+        KS test statistic, either D, D+ or D-.
```
[the remark this answers is not in the document]
Comment (pvanmulbregt) [reply scipy/stats/stats.py:5764]: Done.

File: scipy/stats/stats.py [hunk 15]
```diff
@@ -5819,7 +5689,185 @@ def chisquare(f_obs, f_exp=None, ddof=0, axis=0):
...
+    random variable against a given distribution G(x). Under the null
+    hypothesis, the two distributions are identical, F(x)=G(x). The
+    alternative hypothesis can be either 'two-sided' (default), 'less'
+    or 'greater'. The KS test is only valid for continuous distributions.
+
+    Parameters
+    ----------
+    x : array_like
+        a 1-D array of observations of random variables.
+    cdf : callable
```
[the remark this answers is not in the document]
Comment (pvanmulbregt) [reply scipy/stats/stats.py:5741]: Right now `ks_1samp` doesn't accept a string.  `kstest(x, ...)` does, and it converts the string to a cdf function before passing down to `ks_1samp`.  Since `ks_1samp` is not in the API currently, it doesn't need to support all the myriad combinations of `kstest`.
If one wants the convenience of strings, use the wrapper `kstest` which figures out everything.

File: scipy/stats/stats.py [hunk 16]
```diff
@@ -5819,7 +5689,185 @@ def chisquare(f_obs, f_exp=None, ddof=0, axis=0):
...
+
+    The above lines are equivalent to:
+
+    >>> np.random.seed(987654321)
+    >>> stats.kstest(stats.norm.rvs(size=100), 'norm')
+    (0.058352892479417884, 0.8653960860778898)
+
+    *Test against one-sided alternative hypothesis*
+
+    Shift distribution to larger values, so that ``cdf_dgp(x) < norm.cdf(x)``:
```
[the remark this answers is not in the document]
Comment (pvanmulbregt) [reply scipy/stats/stats.py:5799]: Weird.  `cdf_dgp` has been there for at least 8 years and I have no idea what it means.  Removed it and replaced with `CDF(x) < norm.cdf(x)`

File: scipy/stats/stats.py (hunk 7 above)
Comment (pvanmulbregt) [reply scipy/stats/stats.py:5811]: Removed `asymp` in the example below.

File: scipy/stats/stats.py (hunk 9 above)
Comment (pvanmulbregt) [reply scipy/stats/stats.py:5816]: I believe that the docstring for this function was copied verbatim from the docstriong for the existing `stats.kstest`. If you have replacement doc examples you prefer now I' be happy to use them! If not, I'd like to make those changes in a different PR, as it will likely affect multiple functions.

File: scipy/stats/stats.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (chrisb83) [reply scipy/stats/stats.py]: thanks, when I did my first review, I did not see that these PEP8 issues already were in the existing documentation

**Comment by chrisb83** [comment]
> Personally I'd be in favor of dropping support for these cases. Nobody jumped to explain the use cases after #11710 (comment) and they do make the API complicated and difficult to maintain.

I agree.

File: scipy/stats/tests/test_mstats_basic.py [hunk 17]
```diff
@@ -1436,6 +1436,60 @@ def test_obrientransform(self):
...
+                for alternative in ['less', 'greater', 'two-sided']:
+                    for n in self.get_n():
+                        x, y, xm, ym = self.generate_xy_sample(n)
+                        res1 = stats.ks_1samp(x, stats.norm.cdf, alternative=alternative, mode=mode)
+                        res2 = stats.mstats.ks_1samp(xm, stats.norm.cdf, alternative=alternative, mode=mode)
+                        assert_equal(np.asarray(res1), np.asarray(res2))
+                        res3 = stats.ks_1samp(xm, stats.norm.cdf, alternative=alternative, mode=mode)
+                        assert_equal(np.asarray(res1), np.asarray(res3))
+
+    def test_kstest_1samp(self):
```
Comment (chrisb83) [inline scipy/stats/tests/test_mstats_basic.py:1451]: is this the same test as in line 1439 (copy-paste error?)

**Review by chrisb83 — COMMENTED** [review]
I started to look at the testing: the main focus seems to be on the consistency of the various functions on stats and mstats. I would like to checka few things for the new ks_1samp function, mainly whether the alternatives are correctly implemented before concluding this review.

**Comment by tylerjereddy** [comment]
@chrisb83 @pvanmulbregt Just a reminder we'd like to branch 1.5.x in about 6 days--looks like there's still plenty of activity here so maybe it will converge on time?

**Comment by pvanmulbregt** [comment]
* Added comments to the K-S test functions to make clearer their purpose.

* Dropped support for the mostly undocumented cases which served to only make testing `kstest` use slightly fewer characters but had little real-world applicability and made maintenance more difficult. These include:

```
kstest("norm", "", args=..., N=...)
kstest("norm", False, args=..., N=...)
kstest("norm", [], args=..., N=...)
kstest("norm", 0, args=..., N=...)
```

**Comment by pvanmulbregt** [comment]
@chrisb83 Updated based upon your latest comments. Did I miss any?

**Comment by chrisb83** [comment]
@pvanmulbregt yes, thanks, my comments above have been addressed. 

I took a look at the testing and I think there should be two basic test cases for `ks_1samp`:

- check names of attributes (copy&paste from TestKSTwoSamples)
- one test of the exact option and the alternatives, just to be sure that this is correctly implemented. I generated values for comparison in R below (R version 3.4.1 (2017-06-30))

```
x <- c(-1.23, 0.06, -0.60, 0.17, 0.66, -0.17, -0.08, 0.27, -0.98, -0.99)

res <- ks.test(x, "pnorm", exact = TRUE)
res['statistic'] # 0.2935801
res['p.value'] # 0.2934085

res <- ks.test(x, "pnorm", alternative = 'greater', exact = TRUE)
res['statistic'] # 0.2935801 (D+)
res['p.value'] # 0.1469888

res <- ks.test(x, "pnorm", alternative = 'less', exact = TRUE)
res['statistic'] # 0.1093486 (D-)
res['p.value'] # 0.7327689
```

I don't think more testing is required since there is a lot of testing of  `stats.ksone`. Let me know if you think my suggestion makes sense.

Apart from this point, this PR is ready to be merged in my view.

One follow-up PR could enhance the documentation (e.g. examples). The null hypothesis of `kstest` is incorrectly described as 

> Under the null hypothesis, the two distributions are identical, F(x)=G(x). The alternative hypothesis can be either ‘two-sided’ (default), ‘less’ or ‘greater’. 

However, the null varies depending on alternative. These points are not related to this PR.

**Comment by pvanmulbregt** [comment]
The single CI failure is due to `tools/lint_diff.py`. However the errors and lines mentioned don't seem to match up with the actual code.

```
$ tools/lint_diff.py
./scipy/stats/tests/test_stats.py:2908:27: E231 missing whitespace after ','
./scipy/stats/tests/test_stats.py:2908:29: E231 missing whitespace after ','
./scipy/stats/tests/test_stats.py:2911:28: E231 missing whitespace after ','
./scipy/stats/tests/test_stats.py:2911:31: E231 missing whitespace after ','
./scipy/stats/tests/test_stats.py:2938:27: E231 missing whitespace after ','
./scipy/stats/tests/test_stats.py:2938:29: E231 missing whitespace after ','
./scipy/stats/tests/test_stats.py:2941:28: E231 missing whitespace after ','
./scipy/stats/tests/test_stats.py:2941:31: E231 missing whitespace after ','
8       E231 missing whitespace after ','
```
E.g. test_stats.py:2908-2911 looks fine 
https://github.com/scipy/scipy/blob/e663f2abff06033020e6d069a67cf67d2339793e/scipy/stats/tests/test_stats.py#L2908-L2911

@person142  This branch was created before` lint_diff.py` existed.  Does it have any difficulties in that situation?

**Comment by pvanmulbregt** [comment]
@chrisb83  I added the comparison with R's values from above and took a look at the testing code. 
I suspect that some of the testing is redundant,  now that `stats` and `mstats` have the same underlying implementation. 
The testing of `kstest` can be simplified to test just what it does in terms of wrapping up `ks_1samp` and `ks_2samp`, and leave the heavy lifting to the testing of `ks_1samp` and `ks_2samp`. 
As you point out, the testing of `scipy.stats.stats.ks_1samp` and `scipy.stats.ksone` likely have overlap, or could be partitioned differently.

I think the current status is:
* The functionality is complete, the implementation is final.
* The test suite covers what needs to be tested.
* The testing of `kstest`, `ks_1samp`, `ksone` may test more than is needed and could be rationalized.
* The existing documentation could use some review.

My preference would be to place any changes to the testing/documentation in a new PR after 1.5, as that will take some thinking and review and won't happen before the 1.5 fork.  
And to merge this PR in to make it available for 1.5 (once the `lint_diff.py` issue is understood.)

File: scipy/stats/tests/test_stats.py [hunk 18]
```diff
@@ -2875,41 +2876,112 @@ def test_friedmanchisquare():
...
+    def test_namedtuple_attributes(self):
+        x = np.linspace(-1, 1, 9)
+        # test for namedtuple attribute results
+        attributes = ('statistic', 'pvalue')
+        res = stats.ks_1samp(x, stats.norm.cdf)
+        check_named_results(res, attributes)
+
+    def test_agree_with_r(self):
+        # comparing with some values from R
+        x = np.linspace(-1,1,9)
```
Comment (chrisb83) [inline scipy/stats/tests/test_stats.py:2930]: do the missing spaces after , cause the whitespace error? there are a few more cases above and below

**Comment by chrisb83** [comment]
> My preference would be to place any changes to the testing/documentation in a new PR after 1.5, as that will take some thinking and review and won't happen before the 1.5 fork.

yes, let's get this PR merged before the 1.5 fork. My comments above were addressed, I will just do one final iteration this weekend.

File: scipy/stats/mstats_basic.py [hunk 19]
```diff
@@ -1282,34 +1300,35 @@ def ks_twosamp(data1, data2, alternative="two-sided"):
...
+       alternative.lower()[0], alternative)
+    return scipy.stats.stats.ks_2samp(x, y, alternative=alternative, mode=mode)
+
 
-    return (d, prob)
+ks_twosamp = ks_2samp
 
 
-ks_2samp = ks_twosamp
+def kstest(data1, data2, args=(), alternative='two-sided', mode='auto'):
```
Comment (chrisb83) [inline scipy/stats/mstats_basic.py:1314]: in ks_2samp / ks_twosamp, data1 and data2 are replaced by x and y whereas kstest still uses data1 and data2. is that intendend?

Comment (chrisb83) [reply scipy/stats/mstats_basic.py:1314]: the change in ks_twosamp could even be a problem with backwards compatability?

File: scipy/stats/stats.py [hunk 20]
```diff
@@ -6056,7 +6147,7 @@ def ks_2samp(data1, data2, alternative='two-sided', mode='auto'):
 
     Parameters
     ----------
-    data1, data2 : sequence of 1-D ndarrays
+    x, y : sequence of 1-D ndarrays
```
Comment (chrisb83) [inline scipy/stats/stats.py:6150]: same question as above: can we rename these parameters?

**Comment by chrisb83** [comment]
Two more observation:

- the parameters data1 / x and data2 / y are used in the code and if I am not mistaken, the renaming could create an issue with backwards compatability
- the code now returns a named tuple KStestresult for all variations of the test. can we do such a change in view of the existing functionality that is in use?

**Comment by chrisb83** [comment]
Last but not least: let's not forget to link the issues collected in #9797 that will be closed by this PR

File: scipy/stats/mstats_basic.py (hunk 19 above)
Comment (pvanmulbregt) [reply scipy/stats/mstats_basic.py:1314]: Good spot.  The arguments could be provided by name to `ks_2samp`, even though they are required args.  In `kstest` it is not an issue as `kstest` was not exposed before now.

File: scipy/stats/stats.py (hunk 20 above)
Comment (pvanmulbregt) [reply scipy/stats/stats.py:6150]: I'll change them back in `stats.ks_2samp`.
But that raises an interesting question on `kstest`.
The signature of `kstest` changed:
```
def kstest(rvs, cdf, args=(), N=20, alternative='two-sided', mode='approx')`
-->
def kstest(data1, data2, args=(), N=20, alternative='two-sided', mode='auto')
```

The default `mode` should not be an issue.  But the name change of `rvs, cdf` -> `data1, data2` might be.  The changed names reflect the enhanced nature of the function.  It supports the old use cases of `kstest`, which handled one sample testing.  It also supports two sample testing via `kstest(rvs1, rvs2)`, in which case `rvs2` is definitely not a cdf of any kind.

@rgommers @tylerjereddy Is it permitted to change the names of positional arguments of a public API function?
I did a quick Google search for "kstest(rvs=..." and found a couple of examples specifying `rvs` by keyword, so my assumption is that changing the argument names at this time is not the thing to do.  Just document very clearly that `cdf` can be more than just a `cdf`. Is that correct?

**Comment by WarrenWeckesser** [comment]
>  Is it permitted to change the names of positional arguments of a public API function?

No, that's a backwards incompatible API change.

**Comment by pvanmulbregt** [comment]
Restored original names for backwards compatibility.

**Comment by pvanmulbregt** [comment]
Closes gh-8748 gh-8341 gh-9797

**Review by chrisb83 — APPROVED** [review]
All my comments have been addressed.

**Comment by tylerjereddy** [comment]
CI all green & core dev approval, merging.

**Comment by tylerjereddy** [comment]
I've manually closed & milestoned the cross-linked issues above

**Comment by tylerjereddy** [comment]
Maybe suggest a release note entry directly in the WIP PR: https://github.com/scipy/scipy/pull/12221

**Comment by pvanmulbregt** [comment]
I updated the Wiki with an entry for `kstest`.
